src/Analysis/Ast/Impl/Indexing/SymbolIndex.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Analysis.Indexing {
+    internal sealed class SymbolIndex : ISymbolIndex {
+        private readonly ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>> _index = new ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>>();
+        private bool _empty;
+        
+        public SymbolIndex() {
+            _empty = false;
+        }
+        
+        public bool isNotEmpty() {
+            return _empty;
+        }
+
+
+        public IEnumerable<HierarchicalSymbol> HierarchicalDocumentSymbols(Uri uri)
+            => _index.TryGetValue(uri, out var list) ? list : Enumerable.Empty<HierarchicalSymbol>();
+
+        public IEnumerable<FlatSymbol> WorkspaceSymbols(string query) {
+            foreach (var kvp in _index) {
+                foreach (var found in WorkspaceSymbolsQuery(query, kvp.Key, kvp.Value)) {
+                    yield return found;
```

Comment:
`_index.SelectMany(kvp => WorkspaceSymbolsQuery(query, kvp.Key, kvp.Value))`

We should discuss concurrency here.

Reply:
Discussed offline yesterday